pkg/proxy/proxy.go

```diff
@@ -136,130 +145,191 @@ func (p *Proxy) Run(stopCh <-chan struct{}) (<-chan struct{}, error) {
+		if err != nil {
+			if !p.config.TokenReview {
+				p.handleError(rw, req, errUnauthorized)
+				return
+			}
+
+			// Attempt to passthrough request if valid token
+			if p.reviewToken(rw, req) {
+				// Set no impersonation headers and re-add removed headers.
+				req = req.WithContext(context.WithNoImpersonation(req.Context()))
+
+				handler.ServeHTTP(rw, req)
+				return
+			}
+
+			// Token review failed so error
+			p.handleError(rw, req, errUnauthorized)
+			return
+		}
 
-	klog.V(4).Infof("authenticated request: %s", reqCpy.RemoteAddr)
+		// Failed authorization
+		if !ok {
+			p.handleError(rw, req, errUnauthorized)
+			return
+		}
 
-	// if we have disabled impersonation we can forward the request right away
-	if p.options.DisableImpersonation {
-		klog.V(2).Infof("passing on request with no impersonation: %s", reqCpy.RemoteAddr)
-		// Send original copy here with auth header intact
-		return p.noAuthClientTransport.RoundTrip(req)
-	}
+		klog.V(4).Infof("authenticated request: %s", req.RemoteAddr)
 
-	// check for incoming impersonation headers and reject if any exists
-	if p.hasImpersonation(reqCpy.Header) {
-		return nil, errImpersonateHeader
-	}
+		// Add the user info to the request context
+		req = req.WithContext(genericapirequest.WithUser(req.Context(), info.User))
+		handler.ServeHTTP(rw, req)
+	})
+}
 
-	user := info.User
+// withImpersonateRequest adds the impersonation request handler to the chain.
+func (p *Proxy) withImpersonateRequest(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		// If no impersonation has already been set, return early
+		if context.NoImpersonation(req.Context()) {
+			handler.ServeHTTP(rw, req)
+			return
+		}
 
-	// no name available so reject request
-	if user.GetName() == "" {
-		return nil, errNoName
-	}
+		// If we have disabled impersonation we can forward the request right away
+		if p.config.DisableImpersonation {
+			klog.V(2).Infof("passing on request with no impersonation: %s", req.RemoteAddr)
+			// Indicate we need to not use impersonation.
+			req = req.WithContext(context.WithNoImpersonation(req.Context()))
+			handler.ServeHTTP(rw, req)
+			return
+		}
 
-	// ensure group contains allauthenticated builtin
-	found := false
-	groups := user.GetGroups()
-	for _, elem := range groups {
-		if elem == authuser.AllAuthenticated {
-			found = true
-			break
+		if p.hasImpersonation(req.Header) {
+			p.handleError(rw, req, errImpersonateHeader)
+			return
 		}
-	}
-	if !found {
-		groups = append(groups, authuser.AllAuthenticated)
-	}
 
-	extra := user.GetExtra()
+		user, ok := genericapirequest.UserFrom(req.Context())
+		// No name available so reject request
+		if !ok || len(user.GetName()) == 0 {
+			p.handleError(rw, req, errNoName)
+			return
+		}
 
-	if extra == nil {
-		extra = make(map[string][]string)
-	}
+		// Ensure group contains allauthenticated builtin
```

Comment:
Why do we check for allAuthenticated? I might be reviewing unrelated code that has been moved around here, but it does remind me of the fact that all Google customers are able to *authenticate* with all GKE clusters (authorization/RBAC is used to then not permit people doing things).

That means that `joebloggs@example.com` is able to *authenticate* - does that mean they will have the `allauthenticated` builtin? I can't tell from reading this area of the code exactly why we check for this/what it's for :)

Reply:
This is to ensure that all authenticated requests have the group AllAuthenticated group appended, as per what happens in the API server.

Reply:
This is important since the API server wont appended it to the user info itself during the impersonation loop